plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts

```diff
@@ -0,0 +1,248 @@
+import { getRPCMetadata, RPCType } from '@opentelemetry/core';
+import {
+  InstrumentationBase,
+  InstrumentationConfig,
+  InstrumentationNodeModuleDefinition,
+  safeExecuteInTheMiddle,
+} from '@opentelemetry/instrumentation';
+import { SemanticAttributes } from '@opentelemetry/semantic-conventions';
+import type { HookHandlerDoneFunction } from 'fastify/types/hooks';
+import type { FastifyInstance } from 'fastify/types/instance';
+import type { FastifyReply } from 'fastify/types/reply';
+import type { FastifyRequest } from 'fastify/types/request';
+import {
+  AttributeNames,
+  FastifyNames,
+  FastifyTypes,
+} from './enums/AttributeNames';
+import type { HandlerOriginal } from './types';
+import { safeExecuteInTheMiddleMaybePromise } from './utils';
+import { VERSION } from './version';
+
+export const ANONYMOUS_NAME = 'anonymous';
+
+/** Fastify instrumentation for OpenTelemetry */
+export class FastifyInstrumentation extends InstrumentationBase {
+  constructor(config: InstrumentationConfig = {}) {
+    super(
+      '@opentelemetry/instrumentation-fastify',
+      VERSION,
+      Object.assign({}, config)
+    );
+  }
+
+  init() {
+    return [
+      new InstrumentationNodeModuleDefinition<any>(
+        'fastify',
+        ['^3.0.0'],
+        (moduleExports, moduleVersion) => {
+          diag.debug(`Applying patch for fastify@${moduleVersion}`);
+          return this._patchConstructor(moduleExports);
+        },
+        (moduleExports, moduleVersion) => {
+          diag.debug(`Removing patch for fastify@${moduleVersion}`);
+        }
+      ),
+    ];
+  }
+
+  private _hookOnRequest() {
+    const instrumentation = this;
+    return function onRequest(
+      request: FastifyRequest,
+      reply: FastifyReply,
+      done: HookHandlerDoneFunction
+    ) {
+      if (!instrumentation.isEnabled()) {
+        return done();
+      }
+      const rpcMetadata = getRPCMetadata(context.active());
+      const routeName = request.routerPath;
+      if (routeName && rpcMetadata?.type === RPCType.HTTP) {
+        rpcMetadata.span.updateName(`${request.method} ${routeName || '/'}`);
+      }
+      done();
+    };
+  }
+
+  private _wrapHandler(
+    pluginName: string,
+    hookName: string,
+    original: () => Promise<any> | void
+  ): () => Promise<any> | void {
+    const instrumentation = this;
+    return function (this: any, ...args): Promise<any> | void {
+      if (!instrumentation.isEnabled()) {
+        return original.apply(this, args);
+      }
+
+      const spanName = `${FastifyNames.MIDDLEWARE} - ${
+        original.name || ANONYMOUS_NAME
+      }`;
+
+      const span = instrumentation._startSpan(spanName, {
+        [AttributeNames.FASTIFY_TYPE]: FastifyTypes.MIDDLEWARE,
+        [AttributeNames.PLUGIN_NAME]: pluginName,
+        [AttributeNames.HOOK_NAME]: hookName,
+      });
+
+      return safeExecuteInTheMiddleMaybePromise(
+        () => {
+          return original.apply(this, args);
```

Comment:
no as then it will have wrong context, you can check unit tests they are covering that, after yr suggestions tests are failing as spans are no longer connected

Reply:
For example:
```js
import fastify from 'fastify';
const app = fastify({logger: true});
app.addHook('onRequest', (req, reply, done) => {
    trace.getTracerProvider().getTracer('example').startSpan('span inside fastify middleware').end();
    done();
})
app.get('/', (req, reply) => {
    reply.send('hello world');
});
app.listen(3000, (err, addr) => {
    console.log('listing', {err});
});

```

The manual span which is started inside the middleware function should be the child of the middleware span, but it's not. It's the child of the HTTP.
Tests should be updated as well to reflect this.

I realize it can be a bit of a complex fix, and understand if you choose not to apply it.
I can also offer my assistant in fixing it if you like.

Reply:
fixed